torch/quantization/convert_modules.py

```diff
@@ -0,0 +1,240 @@
+    in swapModule function
+    """
+    def __init__(self, qconfig):
+        super(AbstractQuant, self).__init__()
+        self.config = qconfig['activation']['config']
+        self.add_module('observer', qconfig['activation']['observer'](self.config))
+
+    def forward(self, x):
+        self.observer(x)
+        return x
+
+class AbstractDeQuant(nn.Module):
+    r"""Abstract dequantize module which will be replace to actual DeQuantize module
+    in swapModule function
+    """
+    def __init__(self):
+        super(AbstractDeQuant, self).__init__()
+
+    def forward(self, x):
+        return x
+
+def calculateQParams(MyModule):
+    r""" Calculates Quantizer parameters for activation tensors based on observer statistics
+    Updates the qparams parameter for each module with the quantizer params.
+
+    Args:
+        MyModule: Model with observer stats
+    """
+    if hasattr(MyModule, 'observer'):
+        if isinstance(MyModule.observer, Observer):
+            # Simple symmetric quantization over entire observed range
+            MyModule.qparams = MyModule.observer.calculate_q_params()
+    for Module in MyModule.children():
+        calculateQParams(Module)
+
+def quantize(module, qConfigDict, eval_fn, *eval_args):
+    r"""Given a float module and qConfigDict, convert it to a quantized module
+
+    First it will prepare the module to add observer and abstract quant
+    deqaunt modules, then it calls `eval_fn` which will run the calibration
+    step and observers will record the stats of the tensors, after that we will
+    call `convert` which will convert the module to a quantized module.
+    """
+    prepare(module, qConfigDict)
+    eval_fn(module, *eval_args)
+    convert(module)
+    return
+
+def convert(module):
+    r""" Utility function to traverse module tree and swap modules marked with
+         AbstractQuant and AbstractDeQuant with
+         quantized versions.
+    """
+    calculateQParams(module)
+    return convertToQuantized(module)
+
+def convertToQuantized(module):
+    r"""Convert a module with qparams to a quantized version of the module
+    """
+    module_swapped = swapModule(module)
+    if len(list(module.named_children())) == 0:
+        return module
+
+    reassign = {}
+    for name, mod in module.named_children():
+        new_mod = convertToQuantized(mod)
+        if new_mod is not mod:
+            reassign[name] = new_mod
+
+    for name, mod in reassign.items():
+        setattr(module_swapped, name, mod)
+
+    return module_swapped
+
+def swapModule(mod):
+    r""" Check if a module has a quantized counterpart and swap it.
+    """
+    new_mod = mod
+    if hasattr(mod, 'observer'):
+        if type(mod) == torch.nn.Linear:
+            # Obtain quantized weights and biases
+            observerConfig = mod.qConfig['weight']['config']
+            weight_observer = mod.qConfig['weight']['observer'](observerConfig)
+            quantized_weight, quantized_bias = quantizeWeightAndBias(
```

Comment:
why this logic can't sit inside nnq.Linear constructor? it'd be much more sensible. Is there a case where we want to create a raw nnq.Linear? I doubt so. Then you don't need all this set_state business

Reply:
Yes, we may want to create a raw nnq.Linear. This is important as someone may want to quantize modules their own way (not using this framework) and still be able to access the quantized kernels. This way of splitting the functionality allows for that.